src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,349 @@
+                static bool IsAccessible(Accessibility declaredAccessibility) =>
+                    declaredAccessibility >= Accessibility.Internal || declaredAccessibility == Accessibility.NotApplicable;
+            }
+        }
+
+        private static string ConcatNamespace(string containingNamespace, string name)
+        {
+            Debug.Assert(name != null);
+            if (string.IsNullOrEmpty(containingNamespace))
+            {
+                return name;
+            }
+
+            return containingNamespace + "." + name;
+        }
+
+        private readonly struct TypeOverloadInfo
+        {
+            public TypeOverloadInfo(INamedTypeSymbol nonGenericOverload, INamedTypeSymbol bestGenericOverload, bool containsPublicGenericOverload)
+            {
+                NonGenericOverload = nonGenericOverload;
+                BestGenericOverload = bestGenericOverload;
+                ContainsPublicGenericOverload = containsPublicGenericOverload;
+            }
+
+            public INamedTypeSymbol NonGenericOverload { get; }
+
+            // Generic with fewest type parameters is considered best symbol to show in description.
+            public INamedTypeSymbol BestGenericOverload { get; }
+
+            public bool ContainsPublicGenericOverload { get; }
+
+            public TypeOverloadInfo Aggregate(INamedTypeSymbol type)
+            {
+                if (type.Arity == 0)
+                {
+                    return new TypeOverloadInfo(type, BestGenericOverload, ContainsPublicGenericOverload);
+                }
+
+                var containsPublic = type.DeclaredAccessibility >= Accessibility.Public || ContainsPublicGenericOverload;
+
+                // We consider generic with fewer type parameters better symbol to show in description.
+                if (BestGenericOverload == null || type.Arity < BestGenericOverload.Arity)
+                {
+                    return new TypeOverloadInfo(NonGenericOverload, type, containsPublic);
+                }
+
+                return new TypeOverloadInfo(NonGenericOverload, BestGenericOverload, containsPublic);
+            }
+        }
+
+        private readonly struct ReferenceCacheEntry
+        {
+            public ReferenceCacheEntry(
+                Checksum checksum,
+                ImmutableArray<CompletionItemEntry> cachedItems)
+            {
+                Checksum = checksum;
+                CachedItems = cachedItems;
+            }
+
+            public Checksum Checksum { get; }
+
+            public ImmutableArray<CompletionItemEntry> CachedItems { get; }
+        }
+
+        private readonly struct CompletionItemEntry
+        {
+            public CompletionItemEntry(CompletionItem item, bool isPublic)
+            {
+                Item = item;
+                IsPublic = isPublic;
+            }
+
+            public CompletionItem Item { get; }
+
+            public bool IsPublic { get; }
+        }
+    }
```

Comment:
this is all very nice and well encapsulated.  nice job.

Reply:
:)